common/sgx/endorsements.c

```diff
@@ -0,0 +1,378 @@
+    OE_CHECK(oe_memcpy_s(
+        buffer, size, qe_id_info->issuer_chain, qe_id_info->issuer_chain_size));
+    buffer += qe_id_info->issuer_chain_size;
+
+    // Copy creation datetime
+    OE_CHECK(
+        oe_memcpy_s(buffer, size, creation_datetime, CREATION_DATETIME_SIZE));
+    buffer += CREATION_DATETIME_SIZE;
+
+    if (buffer != (endorsements->buffer + endorsements->buffer_size))
+        OE_RAISE_MSG(
+            OE_FAILURE,
+            "Encountered size mismatch when creating SGX endorsements.");
+
+    *endorsements_buffer = endorsements;
+    *endorsements_buffer_size = size;
+
+    result = OE_OK;
+
+done:
+    if ((result != OE_OK) && endorsements)
+        oe_free(endorsements);
+
+    OE_TRACE_INFO(
+        "Exit call %s: %d(%s)\n", __FUNCTION__, result, oe_result_str(result));
+
+    return result;
+}
+
+/**
+ * Convert a oe_endorsement_t structure to a SGX endorsement structure
+ * (oe_sgx_endorsements_t).
+ *
+ * @param[in] endorsements The endorsements in raw format (oe_endorsements_t)
+ * @param[out] sgx_endorsements The parsed SGX endorsements.
+ */
+oe_result_t oe_parse_sgx_endorsements(
+    const oe_endorsements_t* endorsements,
+    const size_t endorsements_size,
+    oe_sgx_endorsements_t* sgx_endorsements)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    uint32_t* offsets = NULL;
+    uint32_t offsets_size = 0;
+    uint8_t* data_ptr_start = NULL;
+    uint32_t data_size;
+
+    if (endorsements == NULL || sgx_endorsements == NULL)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    // Verify version and enclave type
+    if ((endorsements->version != OE_ATTESTATION_ENDORSEMENT_VERSION) ||
+        (endorsements->enclave_type != OE_ENCLAVE_TYPE_SGX))
+        OE_RAISE_MSG(
+            OE_INVALID_PARAMETER,
+            "Failed to parse SGX endorsement. Invalid version or enclave "
+            "type.");
+
+    if (endorsements->num_elements != OE_SGX_ENDORSEMENT_COUNT)
+        OE_RAISE_MSG(
+            OE_INVALID_PARAMETER,
+            "Failed to parse SGX endorsement. Exepected %d items, but got %d.",
+            OE_SGX_ENDORSEMENT_COUNT,
+            endorsements->num_elements);
+
+    if (endorsements_size > OE_ATTESTATION_ENDORSEMENT_MAX_SIZE ||
+        endorsements->buffer_size > endorsements_size)
+        OE_RAISE_MSG(
+            OE_INVALID_PARAMETER, "Endorsement buffer size is invalid.");
+
+    // Sanity check
+    offsets_size = endorsements->num_elements * sizeof(uint32_t);
+    if (endorsements->buffer_size <= offsets_size)
+        OE_RAISE_MSG(
+            OE_INVALID_PARAMETER, "Endorsement buffer size is invalid.");
+    data_size = endorsements->buffer_size - offsets_size;
+    offsets = (uint32_t*)endorsements->buffer;
+    data_ptr_start = (uint8_t*)(endorsements->buffer + offsets_size);
+
+    memset(sgx_endorsements, 0, sizeof(oe_sgx_endorsements_t));
+
+    for (int i = 0; i < OE_SGX_ENDORSEMENT_COUNT; i++)
+    {
+        uint8_t* item_ptr = data_ptr_start + offsets[i];
```

Comment:
Is this function only ever called by the original enclave?  Or can it be called by the relying party / verifier?  If the latter, this is not safe, since the receiver could be on a big-endian platform/architecture whereas the SGX enclave is little-endian.   So using offsets[i] is not safe without going through an ntohl() check.

Ok to handle this in a future PR if you file a bug to track.

Reply:
Is this also an issue with the current oe_report_t?

Reply:
Haven't looked but I would assume yes it's already a problem, which is why I'm fine with handling it in a separate issue/PR.

Reply:
I just verified via code inspection, and filed issue #2266 to track this.

Reply:
Thank you Dave.